packages/datadog-shimmer/src/shimmer.js

```diff
@@ -44,54 +82,112 @@ function wrapFunction (original, wrapper) {
     if (descriptor.configurable && descriptor.enumerable) {
       target[name] = wrapped
       return target
     }
     descriptor.value = wrapped
   } else {
-    if (descriptor.get || descriptor.set) {
-      // TODO(BridgeAR): What happens in case there is a setter? This seems wrong?
-      // What happens in case the user does indeed set this to a different value?
-      // In that case the getter would potentially return the wrong value?
-      descriptor.get = () => wrapped
+    if (descriptor.get) {
+      // `replaceGetter` may only be used when the getter has no side effect.
+      if (options?.replaceGetter) {
+        descriptor.get = () => wrapped
+      } else {
+        descriptor.get = wrapped
+      }
     } else {
       descriptor.value = wrapped
     }
 
     if (descriptor.configurable === false) {
-      // TODO(BridgeAR): Bail out instead (throw). It is unclear if the newly
-      // created object is actually used. If it's not used, the wrapping would
-      // have had no effect without noticing. It is also unclear what would happen
-      // in case user code would check for properties to be own properties. That
-      // would fail with this code. A function being replaced with an object is
-      // also not possible.
-      return Object.create(target, {
-        [name]: descriptor
-      })
+      // TODO(BridgeAR): This currently only works on the most outer part. The
+      // moduleExports object.
+      //
+      // It would be possible to also implement it for non moduleExports objects
+      // by passing through the moduleExports object and the property names that
+      // are accessed. That way it would be possible to redefine the complete
+      // property chain. Example:
+      //
+      // shimmer.wrap(hapi.Server.prototype, 'start', wrapStart)
+      // shimmer.wrap(hapi.Server.prototype, 'ext', wrapExt)
+      //
+      // shimmer.wrap(hapi, 'Server', 'prototype', 'start', wrapStart)
+      // shimmer.wrap(hapi, 'Server', 'prototype', 'ext', wrapExt)
+      //
+      // That would however still not resolve the issue about the user replacing
+      // the return value so that the hook picks up the new hapi moduleExports
+      // object. To safely fix that, we would have to couple the register helper
+      // with this code. That way it would be possible to directly pass through
+      // the entries.
+
+      // In case more than a single property is not configurable and writable,
+      // Just reuse the already created object.
+      let moduleExports = nonConfigurableModuleExports.get(target)
+      if (!moduleExports) {
+        if (typeof target === 'function') {
+          const original = target
+          moduleExports = function (...args) { return original.apply(original, args) }
+          // This is a rare case. Accept the slight performance hit.
+          skipMethods.add(name)
+          copyProperties(target, moduleExports)
+          skipMethods.delete(name)
```

Comment:
If `name` is already in `skipMethods`, it will wrongly be deleted. Only if it's a new name should we perform the add/delete action:

```suggestion
          const notAlreadySkipped = !skipMethods.has(name)
          if (notAlreadySkipped) skipMethods.add(name)
          copyProperties(target, moduleExports)
          if (notAlreadySkipped) skipMethods.delete(name)
```

You could argue that we'll never call `wrap` with `caller`, `arguments`, `name`, or `length`, in which case this will not be an issue. But it feels unstable 🤷

Reply:
That is indeed correct. This is a case that is specifically designed for functions though, so using these properties on a function would shadow the original one.
I guess we can go for the super safe implementation. It's a rare case anyway.

Reply:
👍 Normally in cases like this, where there's no reason to guard against a specific edge case because you're sure it's never going to happen, I like to put a code comment in there to future me/others that this is intentional.

Reply:
I implemented a fix instead of a comment. I would just skip writing the test case for it.